File: backend/schemas/bs.py
from typing import Literal

from pydantic import BaseModel


class BsLegIn(BaseModel):
    strike: float
    type: Literal["call", "put"]
    side: Literal["long", "short"]
    quantity: int = 1
    expiry: str  # ISO date


class BsRequest(BaseModel):
    symbol: str
    legs: list[BsLegIn]
    iv_override: float | None = None
    rate: float | None = None  # if None, server fetches DGS3MO


class BsPoint(BaseModel):
    price: float
    pnl: float


class BsGreeks(BaseModel):
    delta: float
    gamma: float
    theta: float
    vega: float


class BsResponse(BaseModel):
    payoff_at_expiry: list[BsPoint]
    current_value: list[BsPoint]
    cost_debit_credit: float                # negative = credit
    max_gain: float | None                  # None when unlimited_gain
    max_loss: float | None                  # None when unlimited_loss
    unlimited_gain: bool
    unlimited_loss: bool
    edge_pnl_low: float                     # PnL at the leftmost displayed price
    edge_pnl_high: float                    # PnL at the rightmost displayed price
    breakevens: list[float]
    greeks: BsGreeks
    prob_profit: float | None               # P(payoff_at_expiry > 0) from MC

File: frontend/src/types/bs.ts
import { z } from "zod";

export const BsPointSchema = z.object({
  price: z.number(),
  pnl: z.number(),
});

export const BsGreeksSchema = z.object({
  delta: z.number(),
  gamma: z.number(),
  theta: z.number(),
  vega: z.number(),
});

export const BsResponseSchema = z.object({
  payoff_at_expiry: z.array(BsPointSchema),
  current_value: z.array(BsPointSchema),
  cost_debit_credit: z.number(),
  max_gain: z.number().nullable(),
  max_loss: z.number().nullable(),
  unlimited_gain: z.boolean(),
  unlimited_loss: z.boolean(),
  edge_pnl_low: z.number(),
  edge_pnl_high: z.number(),
  breakevens: z.array(z.number()),
  greeks: BsGreeksSchema,
  prob_profit: z.number().nullable(),
});

export type BsResponse = z.infer<typeof BsResponseSchema>;
export type BsPoint = z.infer<typeof BsPointSchema>;
export type BsGreeks = z.infer<typeof BsGreeksSchema>;

export const STRATEGY_TYPES = [
  // Order matters — Long Straddle is the visual default per Phase 5 spec.
  { value: "long_straddle",     label: "Long Straddle (ATM)" },
  { value: "long_call",         label: "Long Call" },
  { value: "long_put",          label: "Long Put" },
  { value: "short_call",        label: "Short Call" },
  { value: "short_put",         label: "Short Put" },
  { value: "bull_call_spread",  label: "Bull Call Spread" },
  { value: "bear_put_spread",   label: "Bear Put Spread" },
  { value: "bull_put_spread",   label: "Bull Put Spread (credit)" },
  { value: "bear_call_spread",  label: "Bear Call Spread (credit)" },
  { value: "long_strangle",     label: "Long Strangle" },
  { value: "iron_condor",       label: "Iron Condor" },
  { value: "calendar_spread",   label: "Calendar Spread" },
] as const;

export type StrategyType = (typeof STRATEGY_TYPES)[number]["value"];
